### ISSUE TITLE
Initial values of form fields in frontend are not visible

### ISSUE THREAD

**Issue by schatimo** [pydantic/FastUI#91]
Hi,

`FastUI` is really a great package, many thanks for providing a first version, it really simplifies the development process and allows to focus on backend developments.

I took the demo provided to test the functionality in the *frontend* "select a user defined form data (`FormUserQuery`), click on a button and then automatically populate the actual form fields (`SelectForm`)". For my "M"WE, please see below.

I do so by calling the `"/get_user_endpoint"` to get the user form data and then setting the initial values of `SelectForm` to the values stored in the user form data (in the endpoint `""`). The issue is that in the frontend I cannot see the initial values for the fields of `SelectForm` for a given user defined form select (only for the field `user_query`, but not for `tool`). However, when looking at the response of the endpoint `"/api"`, the initial values are in the API response and thus, are sent to the frontend. Similarly, when looking into the logs of npm, I also see that the initial values are there and populated in `npm-fastui.src.components.form.FormComp` in the `fieldProps[].initial`.

Though, when *directly* calling the URL, e.g., "http://localhost:3000/example?query_config=%7B%22tool%22%3A%22hammer%22%2C%22user_query%22%3A%22test%22%7D" from the browser (and not from "http://localhost:3000/example" and selecting the user query), then the user specific form data is available in the frontend. (This assumes the router endpoints below have the prefix `example`.)

Has anyone faced a similar issue and has an idea how to solve this issue? As I spent quite some time on investigating this issue, any help is appreciated.

Best,
Timo

```python
from fastapi import APIRouter

from fastui import FastUI, AnyComponent, components as c
from fastui.forms import SelectSearchResponse, FormResponse, fastui_form
from fastui.events import GoToEvent
from shared import demo_page
from enum import StrEnum
from pydantic import Field, BaseModel
from typing import Annotated, List
import json

router = APIRouter()


@router.get("/get_user_queries", response_model=SelectSearchResponse)
def get_fonds() -> SelectSearchResponse:
    user_queries = {"test": {"tool": "hammer"}}
    options = [{"label": k, "value": k} for k, _ in user_queries.items()]
    return SelectSearchResponse(options=options)


class ToolEnum(StrEnum):
    hammer = "hammer"
    screwdriver = "screwdriver"


class SelectForm(BaseModel):
    tool: ToolEnum = Field(title="Select Tool")
    user_query: str | None = Field(title="Enter the name of your query.")


class FormUserQuery(BaseModel):
    query_name: str = Field(
        json_schema_extra={"search_url": "/api/example/get_user_queries"},
    )


@router.post("/get_user_query")
async def get_user_query(
    form: Annotated[FormUserQuery, fastui_form(FormUserQuery)]
) -> FormResponse:
    user_queries = {"test": {"tool": "hammer"}}
    query_config_dict = user_queries[form.query_name]
    query_config_dict["user_query"] = form.query_name
    return FormResponse(
        event=GoToEvent(
            url="",
            query={"query_config": SelectForm(**query_config_dict).model_dump_json()},
        )
    )


@router.post("/post_form_data")
async def post_form_data(form: Annotated[SelectForm, fastui_form(SelectForm)]) -> FormResponse:
    return FormResponse(event=GoToEvent(url="", query={"post_text": f"You selected {form.tool}."}))


@router.get("", response_model=FastUI, response_model_exclude_none=True)
def forms_view(query_config: str | None = None, post_text: str | None = None) -> list[AnyComponent]:
    query_config_dict = None
    if query_config:
        query_config_dict = json.loads(query_config)
    return demo_page(
        c.ModelForm[FormUserQuery](
            submit_url="/api/example/get_user_query"
        ),
        c.ModelForm[SelectForm](
            submit_url="/api/example/post_form_data",
            initial=query_config_dict,
        ),
        c.Text(text=post_text if post_text else "Nothing posted yet."),
        title="Example",
    )
```

**Comment by schatimo** [comment]
The issue with the setup described above is that the `Select` component is not re-rendered when the `defaultValue` prop is updated from the backend via the `initial` prop (e.g., for `FormFieldSelectReactComp`). Therefore, to dynamically set values from the backend, e.g., from saved user queries as described, I propose to take instead of the `defaultValue` prop the `value` prop from the `Select` component which triggers a re-render upon change and adjust the respective components in `FormField` such as `FormFieldSelectSearchProps` and `FormFieldSelectReactComp` to something like:

```typescript
export const FormFieldSelectReactComp: FC<FormFieldSelectProps> = (props) => {
  const { name, required, locked, options, multiple, initial, placeholder, onChange } = props

  const className = useClassName(props)
  const classNameSelectReact = useClassName(props, { el: 'select-react' })
  // Bring initial value in `{label: l, value: v}` structure, otherwise keys are missing leading to warnings
  const initialSelectedValue =
    initial !== undefined && Array.isArray(initial)
      ? initial.map((initialValue: any) => findDefault(options, initialValue))
      : initial
  const [selectedValue, setSelectedValue] = useState<any>(initialSelectedValue)

  // Update the selected value when the initial value changes
  useEffect(() => {
    if (multiple && Array.isArray(initial)) {
      const initialOption = initial.map((initialValue: any) => findDefault(options, initialValue))
      setSelectedValue(initialOption)
    } else {
      const initialOption = findDefault(options, initial)
      setSelectedValue(initialOption)
    }
  }, [initial, options, multiple])

  const reactSelectOnChanged = (selectedOption: any) => {
    // TODO this is a hack to wait for the input to be updated, can we do better?
    setTimeout(() => {
      setSelectedValue(selectedOption)
      onChange && onChange()
    }, 50)
  }

  return (
    <div key={name} className={className}>
      <Label {...props} />
      <Select
        id={inputId(props)}
        classNamePrefix="fastui-react-select"
        className={classNameSelectReact}
        isMulti={multiple ?? false}
        isClearable
        value={selectedValue}
        name={name}
        required={required}
        isDisabled={locked}
        options={options}
        aria-describedby={descId(props)}
        styles={styles}
        placeholder={placeholder}
        onChange={reactSelectOnChanged}
      />
      <ErrorDescription {...props} />
    </div>
  )
}
```

Any views on the proposal?

**Comment by samuelcolvin** [comment]
closed via #161.
